### PR TITLE
30 create classes for rhs metric t

### PR DESCRIPTION
**Pull request by seyhani**

### CODE REVIEW

File: lib/util.hpp [hunk 1]
```diff
@@ -47,11 +47,29 @@ typedef pair<cid_t, cid_t> cid_pair;
...
+    enum class Type { GT = 0, GE, LT, LE, EQ };
+
+    Op();
+    Op(Op::Type op);
+
+    Type get() const;
+
+    friend ostream& operator<<(ostream& os, const Op& op);
+
+    static bool eval(unsigned int lhs_val, const Op& op, unsigned int rhs_val);
```
Comment (seyhani) [inline lib/util.hpp:61]:
```suggestion
```

File: lib/workload.hpp [hunk 2]
```diff
@@ -41,9 +41,72 @@ The Grammar:
...
+//************************************* Constant *************************************//
+
+class Constant : public Rhs {
+
+public:
+    Constant(unsigned int coeff);
+    unsigned int get_coeff() const;
+
+private:
+    unsigned int coeff;
```
Comment (seyhani) [inline lib/workload.hpp:99]: I think `value` is a better name.

File: src/contention_point.cpp [hunk 3]
```diff
@@ -2284,50 +2304,55 @@ bool ContentionPoint::eval_spec(Comp comp, IndexedExample* eg, time_range_t time
...
+        eval_m_expr(m_expr, eg, time, res);
+    } else if (time_expr) {
+        eval_Time(*time_expr, eg, time, res);
+    } else if (const_expr) {
         res.valid = true;
-        res.value = get<unsigned int>(rhs);
+        res.value = const_expr->get_coeff();
     } else {
         cout << "ContentionPoint::eval_rhs: invalid variant." << endl;
     }
```
Comment (seyhani) [inline src/contention_point.cpp:2338]: Do it the same way you have done for `get_expr(MExpr* lhs` (without else).

File: src/shared_config.cpp [hunk 4]
```diff
@@ -80,8 +80,18 @@ unsigned int Dists::input_queue_cnt() {
...
+    auto value = op_dist(gen);
+    Op::Type type;
+    switch (value) {
+        case 0: type = Op::Type::GT; break;
+        case 1: type = Op::Type::GE; break;
+        case 2: type = Op::Type::LT; break;
+        case 3: type = Op::Type::LE; break;
+        case 4: type = Op::Type::EQ; break;
+        default: throw std::invalid_argument("Unknown operation");
+    }
```
Comment (seyhani) [inline src/shared_config.cpp:93]:
```suggestion
    Op::Type type = (Op::Type) op_dist(gen);
```

File: src/spec_factory.cpp [hunk 5]
```diff
@@ -177,109 +179,91 @@ void SpecFactory::pick_neighbors(Comp& spec, vector<Comp>& neighbors) {
...
+Rhs* SpecFactory::random_rhs(RandomSpecGenerationParameters params) {
     unsigned int rhs_type = dists->rhs();
     while (rhs_type == 0) {
         rhs_type = dists->rhs();
     }
 
-    if (params.time_valid && rhs_type == 1) return random_time();
+    if (params.time_valid && rhs_type == 1){
+        Time time = random_time();
+        return dynamic_cast<Rhs*>(new Time(time));
```
Comment (seyhani) [inline src/spec_factory.cpp:194]:
```suggestion
        return new Time(time);
```

File: src/spec_factory.cpp [hunk 6]
```diff
@@ -177,109 +179,91 @@ void SpecFactory::pick_neighbors(Comp& spec, vector<Comp>& neighbors) {
...
                                      RandomSpecGenerationParameters params) {
-    switch (rhs.index()) {
-        // trf
-        case 0: {
-            m_expr_t trf = get<m_expr_t>(rhs);
-            vector<m_expr_t> trf_neighbors;
-            pick_m_expr_neighbors(trf, trf_neighbors);
-            neighbors.insert(neighbors.end(), trf_neighbors.begin(), trf_neighbors.end());
-            break;
+    MExpr* trf = dynamic_cast<MExpr*>(rhs);
```
Comment (seyhani) [inline src/spec_factory.cpp:209]:
```suggestion
    MExpr* mexpr = dynamic_cast<MExpr*>(rhs);
```

File: src/contention_point.cpp (hunk 3 above)
Comment (lucedes27) [reply src/contention_point.cpp]: [Remove else statement from ContentionPoint::eval_rhs](https://github.com/all-things-networking/fperf/pull/32/commits/bc9ae12f80cdc0bc38873d29497a861f15078406)

**Pull request merged**